router/handler_edge_ctrl/apiSessionAdded.go

```diff
@@ -43,15 +64,233 @@ func (h *apiSessionAddedHandler) HandleReceive(msg *channel2.Message, ch channel
 	}()
 }
+
+func (h *apiSessionAddedHandler) startSyncApplier() {
+	for apiSessions := range h.syncReady {
+		for _, apiSession := range apiSessions {
+			h.sm.AddApiSession(apiSession)
+		}
+		h.sm.RemoveMissingApiSessions(apiSessions)
+
+		pfxlog.Logger().Infof("finished sychronizing api sessions [count: %d]", len(apiSessions))
+	}
+}
+
+func (h *apiSessionAddedHandler) startSyncFail() {
+
+	for err := range h.syncFail {
+		pfxlog.Logger().Errorf("failed to synchronize sessions, retrying: %v", err)
+
+		h.syncTracker.Stop()
+		h.syncTracker = nil
+
+		resync := &edge_ctrl_pb.RequestClientReSync{
+			Reason: fmt.Sprintf("error during sync: %v", err),
+		}
+
+		resyncProto, _ := proto.Marshal(resync)
+
+		resyncMsg := channel2.NewMessage(env.RequestClientReSyncType, resyncProto)
+		_ = h.control.Send(resyncMsg)
+	}
+}
+
+func (h *apiSessionAddedHandler) legacySync(reqWithState *apiSessionAddedWithState) {
+	pfxlog.Logger().Warn("using legacy sync logic some connections may be dropped")
+	for _, apiSession := range reqWithState.ApiSessions {
+		h.sm.AddApiSession(apiSession)
+	}
+
+	h.sm.RemoveMissingApiSessions(reqWithState.ApiSessions)
+}
+
+func (h *apiSessionAddedHandler) startRecieveSync() {
+	for reqWithState := range h.reqChan {
+		switch reqWithState.SyncStrategyType {
+		case string(sync_strats.RouterSyncStrategyInstant):
+			h.instantSync(reqWithState)
+		case "":
+			pfxlog.Logger().Warn("syncStrategy is not specifieid, old controller?")
+			h.legacySync(reqWithState)
+		default:
+			pfxlog.Logger().Warnf("syncStrategy [%s] is not supported", reqWithState.SyncStrategyType)
+			h.legacySync(reqWithState)
+		}
+	}
+}
+
+func (h *apiSessionAddedHandler) instantSync(reqWithState *apiSessionAddedWithState) {
+	logger := pfxlog.Logger().WithField("strategy", reqWithState.SyncStrategyType)
+
+	state := &sync_strats.InstantSyncState{}
+
+	if reqWithState.InstantSyncState == nil {
+		logger.Panic("syncState is empty, cannot continue")
+	}
+
+	if reqWithState.InstantSyncState.Id == "" {
+		logger.Panic("syncState id is empty, cannot continue")
+	}
+
+	//if no id or the sync id is newer, reset
+	if h.syncTracker == nil || h.syncTracker.syncId == "" || h.syncTracker.syncId < state.Id {
+		logger.Warnf("new syncId [%s], resetting", state.Id)
+
+		if h.syncTracker != nil {
+			h.syncTracker.Stop()
+		}
+
+		h.syncTracker = newApiSessionSyncTracker(state.Id)
+
+		h.syncTracker.StartDeadline(h.syncReady, h.syncFail, 20*time.Second)
+	}
+
+	//ignore older syncs
+	if h.syncTracker.syncId > state.Id {
+		logger.Warnf("older syncId [%s], ignoring", state.Id)
+		return
+	}
+
+	h.syncTracker.Add(reqWithState)
+
+}
+
+type apiSessionSyncTracker struct {
+	syncId           string
+	syncLastRecieved bool
+	reqsWithState    map[int]*apiSessionAddedWithState
+	hasLast          bool
+	lastSeq          int
+	stop             chan struct{}
+	deadline         sync.Once
+	isDone           bool
+}
+
+func newApiSessionSyncTracker(id string) *apiSessionSyncTracker {
```

Comment:
I am not in love with this or the analogous `sessionSyncTracker`. It feels like a sequencer but I wasn't sure if a full-blown sequencer was necessary. I looked at the sequencer interface in `foundation` and some of the implementations. Still wasn't sure if the use fit.